File: server/rank.go
package boards

import (
	"fmt"

	"roci.dev/fracdex"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

// Fractional ranks, server side.
//
// This is the THIRD implementation of the same key space — the app uses npm
// `fractional-indexing` (tinycld/boards/lib/rank.ts), the CLI uses
// `roci.dev/fracdex` (cli/rank.go), and this uses that same Go port. The
// library choice is not incidental: a rank is computed by whoever inserts the
// row, and now three callers insert cards. If any two disagreed, a card
// created here would sort somewhere the app would not have put it, with no
// error raised anywhere.
//
// fracdex is therefore the first external dependency in this module, whose
// go.mod otherwise documents having none. It needs no go.sum: the hash lands
// in the gitignored go.work.sum, which `go mod download` regenerates — the
// same treatment every other dependency here gets.
//
// The two properties cli/rank.go warns about hold verbatim:
//
//  1. RANKS ARE NOT UNIQUE. Two writers splitting the same gap compute the
//     same string, and there is no unique index on `position`. Every query
//     ordering by rank MUST sort `position, id`.
//  2. THE KEY SPACE IS ASCII-ORDERED and lengths vary. Never parse a rank,
//     compare it numerically, or assume a width.

// rankAppendToList returns the rank placing a card after every card already in
// `listID`.
//
// Reads the maximum position directly rather than loading the list's cards:
// appending needs one string, and a board column can hold thousands of rows.
// The ordering is the string ordering the key space defines, which is what
// SQLite's default (BINARY) collation on a text column already is — the same
// comparison `ORDER BY position, id` makes everywhere else.
//
// An empty list (no rows, or a NULL max) yields the first rank, exactly as
// rankForAppend's empty-slice case does.
// rankAfter is the rank placing a row after `last` (the first rank when
// `last` is empty) — what a new sprint takes to land after the planned ones.
func rankAfter(last string) (string, error) {
	return fracdex.KeyBetween(last, "")
}

// ranksAppending returns n ascending ranks that all sort after `last` (or from
// the start of the key space when it is empty).
//
// The Go counterpart of lib/rank.ts's `ranksAfter`, and needed for the same
// reason the bulk actions needed that one: rankAfter called n times against
// unchanged state returns the SAME string every time, so an importer laying
// down a column of cards would put every one of them on one rank and lose the
// order to the id tiebreaker.
//
// fracdex.NKeysBetween is the direct port of the npm package's
// generateNKeysBetween, which is what makes the two byte-compatible — the
// property this file's header calls structural. testdata/nkeys_vectors.json
// holds keys captured FROM the npm package, so a divergence fails here rather
// than showing up as cards sorting differently in the app than in an import.
func ranksAppending(last string, n int) ([]string, error) {
	if n < 0 {
		return nil, fmt.Errorf("boards: rank count must not be negative (got %d)", n)
	}
	return fracdex.NKeysBetween(last, "", uint(n))
}

func rankAppendToList(app core.App, listID string) (string, error) {
	var last string
	err := app.DB().
		NewQuery("SELECT COALESCE(MAX(position), '') AS p FROM boards_cards WHERE list = {:list}").
		Bind(dbx.Params{"list": listID}).
		Row(&last)
	if err != nil {
		return "", err
	}
	return fracdex.KeyBetween(last, "")
}

File: tinycld/boards/lib/rank.ts
// Fractional ranks for ordering lists and cards.
//
// A rank is an opaque string that sorts lexicographically. Inserting between two
// neighbours produces a new string without touching either of them, so a move
// rewrites ONE row — which is what makes an optimistic drag safe. Renumbering
// integer positions would rewrite every sibling and let two concurrent moves
// clobber each other.
//
// The implementation is `fractional-indexing`, already present in the workspace
// as a TanStack DB dependency. This module exists to name the operations in
// boards' own terms and to keep the choice of implementation in one place — if
// this ever moves into core, callers do not change.
//
// TWO PROPERTIES CALLERS MUST KNOW:
//
// 1. RANKS ARE NOT UNIQUE. Two clients inserting into the same gap while offline
//    compute the same string, and there is deliberately no unique index on
//    `position` — a unique constraint would turn a harmless tie into a failed
//    optimistic insert. Every query ordering by rank must therefore sort
//    `ORDER BY position, id`: `id` is the stable tiebreaker that keeps a tie
//    rendering identically on every client instead of flickering between them.
//
// 2. THE KEY SPACE IS ASCII-ORDERED, not just digits — a rank may contain any of
//    0-9, A-Z, a-z, and its length varies. Never parse a rank, compare it
//    numerically, or assume a fixed width; only ever compare two ranks as
//    strings and hand them back to the functions below.
import { generateKeyBetween, generateNKeysBetween } from 'fractional-indexing'

/** The rank of the first item in an empty list. */
export const FIRST_RANK = generateKeyBetween(null, null)

/**
 * A rank strictly between `before` and `after`.
 *
 * `before === null` means the start of the list, `after === null` the end; both
 * null yields {@link FIRST_RANK}. Throws when the two are equal or out of order,
 * which always means the caller passed unsorted or duplicate neighbours.
 *
 * The ordering check is OURS, not the library's, and it is deliberately
 * redundant with 3.x's own guard. 3.x throws on reversed arguments; 4.0.0
 * silently swaps them and returns a plausible-looking key instead
 * (`generateKeyBetween('a1', 'a0')` → `'a0V'`), which would land a card in the
 * wrong place with no error anywhere — the board just quietly reorders itself.
 * The peer range is pinned `<4` so that version cannot resolve, but a peer range
 * is a request, not an enforcement: it is satisfied by whatever the consuming
 * workspace hoists, and this repo does not own every install. Checking here
 * makes the guarantee independent of which version actually loaded.
 */
export function rankBetween(before: string | null, after: string | null): string {
    if (before !== null && after !== null && before >= after) {
        throw new Error(
            `rank: before must sort strictly before after (got '${before}', '${after}')`
        )
    }
    return generateKeyBetween(before, after)
}

/**
 * `count` ascending ranks — for seeding a board's lists or a card's checklist in
 * one pass, rather than calling {@link rankBetween} in a loop.
 */
export function initialRanks(count: number): string[] {
    if (!Number.isInteger(count) || count < 0) {
        throw new Error(`rank: count must be a non-negative integer (got ${count})`)
    }
    return generateNKeysBetween(null, null, count)
}

/**
 * `count` ascending ranks that all sort AFTER `highest`.
 *
 * A bulk move drops N cards into one column at once, and {@link rankBetween}
 * (or lib/move.ts's `rankForAppend`) called N times against unchanged state
 * returns the SAME string every time — every card would land on one rank and
 * their order would be decided by the `id` tiebreaker rather than by the order
 * the user selected them in.
 *
 * `highest` is the greatest rank already in the target column, or null for an
 * empty one. It is the greatest rather than the last element because under a
 * non-manual sort the caller's column is not in rank order — the same reason
 * `rankForAppend` scans for a maximum instead of reading `cards.at(-1)`.
 */
export function ranksAfter(highest: string | null, count: number): string[] {
    if (!Number.isInteger(count) || count < 0) {
        throw new Error(`rank: count must be a non-negative integer (got ${count})`)
    }
    return generateNKeysBetween(highest, null, count)
}
